src/OpenTelemetry.Api/DistributedContext/DistributedContext.cs

```diff
@@ -33,13 +35,113 @@ namespace OpenTelemetry.Context
+                    foreach (DistributedContextEntry entry in entries)
+                    {
+                        int i;
+                        for (i = 0; i < list.Count; i++)
+                        {
+                            if (entry.Key == list[i].Key)
+                            {
+                                list[i] = entry;
+                                break;
+                            }
+                        }
+
+                        if (i >= list.Count)
+                        {
+                            list.Add(entry);
+                        }
+                    }
+
+                    this.entries = list;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DistributedContext"/> struct.
+        /// </summary>
+        /// <param name="key">The key of the context entry.</param>
+        /// <param name="value">The value of the context entry.</param>
+        public DistributedContext(string key, string value)
+        {
+            if (key is null || value is null)
+            {
+                throw new ArgumentNullException(key is null ? nameof(key) : nameof(value));
+            }
+
+            this.entries = carrier is NoopDistributedContextCarrier ? emptyList : new List<DistributedContextEntry>(1) { new DistributedContextEntry(key, value) };
         }
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DistributedContext"/> struct.
+        /// </summary>
+        /// <param name="entry">The distributed context entry.</param>
+        public DistributedContext(DistributedContextEntry entry)
+        {
+            if (entry.Key is null || entry.Value is null)
+            {
+                throw new ArgumentNullException(entry.Key is null ? nameof(entry.Key) : nameof(entry.Value));
+            }
+
+            this.entries = carrier is NoopDistributedContextCarrier ? emptyList : new List<DistributedContextEntry>(1) { entry };
+        }
+
+        /// <summary>
+        /// Gets empty object of <see cref="DistributedContext"/> struct.
+        /// </summary>
+        public static DistributedContext Empty { get; } = new DistributedContext(emptyList);
+
         /// <summary>
         /// Gets the current <see cref="DistributedContext"/>.
         /// </summary>
-        public static DistributedContext Current { get; /* TODO: to be implemented  */ }
+        public static DistributedContext Current => carrier.Current;
+
+        /// <summary>
+        /// Gets or sets the default carrier instance of the <see cref="DistributedContextCarrier"/> class.
+        /// SDK will need to override the value to AsyncLocalDistributedContextCarrier.Instance.
+        /// </summary>
+        public static DistributedContextCarrier Carrier
+        {
+            get => carrier;
+            set
+            {
+                if (value is null)
```

Comment:
Are you suggesting to have like DistributedContextFactory which create DistributedContext objects and allow setting the carrier there? if so, this will require creating many other abstracting classes and really will complicate it. I am treating DistributedContext as special because we defined it not abstracted at all and self contained.  I am not sure why the factory pattern is better here. please clarify if I am missing any point.

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/issues/373